pkg/outputs/kafka_output/kafka_output.go

```diff
@@ -280,20 +280,61 @@ func (k *kafkaOutput) RegisterMetrics(reg *prometheus.Registry) {
 }
 
 func (k *kafkaOutput) worker(ctx context.Context, idx int, config *sarama.Config) {
-	var producer sarama.SyncProducer
+	var producer sarama.AsyncProducer
 	var err error
 	defer k.wg.Done()
 	workerLogPrefix := fmt.Sprintf("worker-%d", idx)
 	k.logger.Printf("%s starting", workerLogPrefix)
 CRPROD:
-	producer, err = sarama.NewSyncProducer(strings.Split(k.Cfg.Address, ","), config)
+	producer, err = sarama.NewAsyncProducer(strings.Split(k.Cfg.Address, ","), config)
 	if err != nil {
 		k.logger.Printf("%s failed to create kafka producer: %v", workerLogPrefix, err)
 		time.Sleep(k.Cfg.RecoveryWaitTime)
 		goto CRPROD
 	}
 	defer producer.Close()
 	k.logger.Printf("%s initialized kafka producer: %s", workerLogPrefix, k.String())
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case msg, ok := <-producer.Successes():
+				if !ok {
+					return
+				}
+				if k.Cfg.EnableMetrics {
+					start, ok := msg.Metadata.(time.Time)
+					if ok {
+						kafkaSendDuration.WithLabelValues(config.ClientID).Set(float64(time.Since(start).Nanoseconds()))
+					}
+					kafkaNumberOfSentMsgs.WithLabelValues(config.ClientID).Inc()
+					kafkaNumberOfSentBytes.WithLabelValues(config.ClientID).Add(float64(msg.Value.Length()))
+				}
+			}
+		}
+	}()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-producer.Errors():
+				if !ok {
+					return
+				}
+				if k.Cfg.Debug {
+					k.logger.Printf("%s failed to send a kafka msg to topic '%s': %v", workerLogPrefix, err.Msg.Topic, err.Err)
+				}
+				if k.Cfg.EnableMetrics {
+					kafkaNumberOfFailSendMsgs.WithLabelValues(config.ClientID, "send_error").Inc()
+				}
+			}
+		}
+	}()
+
```

Comment:
These 2 goroutines can be just one. Have one `for` loop that selects on the 3 channels: `ctx.Done()`, `producer.Successes()` and `producer.Errors()`

Reply:
Makes sense. Fixed.